examples/Nat/PacketEncapsulation.cs

```diff
@@ -0,0 +1,213 @@
+
+Use of this source code is governed by the Apache 2.0 license; see LICENSE.
+*/
+
+using System;
+using PacketDotNet;
+
+namespace Pax.Examples.Nat
+{
+  /// <summary>
+  /// Class that provides strongly typed access to packets and their encapsulated packets.
+  /// </summary>
+  /// <typeparam name="TTransport">The type of the transport-layer packet</typeparam>
+  /// <typeparam name="TNode">The type of node.</typeparam>
+  public abstract class PacketEncapsulation<TTransport,TNode> where TTransport : Packet where TNode : Node
+  {
+    /// <summary>
+    /// The link-layer packet.
+    /// </summary>
+    public EthernetPacket LinkPacket { get; }
+
+    /// <summary>
+    /// The network-layer packet.
+    /// </summary>
+    public IpPacket NetworkPacket { get; }
+
+    /// <summary>
+    /// The transport-layer packet.
+    /// </summary>
+    public TTransport TransportPacket { get; }
+
+    /// <summary>
+    /// Casts the packet and it's payload packets to the correct types.
+    /// </summary>
+    /// <param name="packet"></param>
+    public PacketEncapsulation(Packet packet)
+    {
+      LinkPacket = (EthernetPacket)packet;
+      NetworkPacket = (IpPacket)LinkPacket.PayloadPacket;
+      TransportPacket = (TTransport)NetworkPacket.PayloadPacket;
+    }
+
+    /// <summary>
+    /// Updates the checksums of the packets for each layer.
+    /// </summary>
+    public void UpdateChecksums()
+    {
+      // Update transport checksum
+      UpdateTransportChecksum();
+
+      // Update IPv4 checksum
+      if (NetworkPacket is IPv4Packet)
+        ((IPv4Packet)NetworkPacket).UpdateIPChecksum();
+    }
+
+    /// <summary>
+    /// Rewrites the packet's source.
+    /// </summary>
+    /// <param name="node">The source node.</param>
+    public void SetSource(TNode node)
+    {
+      // Rewrite the MAC address
+      LinkPacket.SourceHwAddress = node.MacAddress;
+
+      // Rewrite the source IP address
+      NetworkPacket.SourceAddress = node.Address;
+
+      // Allow subclasses to set the transport layer protocol values
+      OnSetSource(node);
+    }
+
+    /// <summary>
+    /// Rewrites the packet's destination.
+    /// </summary>
+    /// <param name="node">The destination node.</param>
+    public void SetDestination(TNode node)
+    {
+      // Rewrite the MAC address
+      LinkPacket.DestinationHwAddress = node.MacAddress;
+
+      // Rewrite the destination IP address
+      NetworkPacket.DestinationAddress = node.Address;
+
+      // Allow subclasses to set the transport layer protocol values
+      OnSetDestination(node);
+    }
+
+    /// <summary>
+    /// Gets the source node of this packet (where it appears to originate from).
+    /// </summary>
+    /// <param name="incomingNetworkInterface">The network interface this packet arrived on.</param>
+    /// <returns>A <see cref="TNode"/> representing the source of the packet.</returns>
+    public abstract TNode GetSourceNode(int incomingNetworkInterface = -1);
```

Comment:
Is `-1` intended to be a rogue value? If not, what does it mean? (Worth adding to the `param` comment if `incomingNetworkInterface` is expected to be nonnegative.)